### DISCUSSION TITLE
What’s the best library to write opening_hours parser?

### DISCUSSION THREAD

**Discussion by strump** [organicmaps/organicmaps#3151] [Q&A]
Currently there are couple of problems with opening_hours parser. It’s written with Boost.Spirit and doesn’t pass all unit tests. There are issues with complex opening_hours cases which OM can’t parse: #1612, #3046, #1366, #1247.

What would be the best library to replace Boost.Spirit? We can use LL parser generators ([Flex](https://en.wikipedia.org/wiki/Flex_(lexical_analyser_generator)), [Carbuetta](https://www.carburetta.com/), Quex). Or write own simple LL parser. Or port from [PHP](https://github.com/ujamii/osm-opening-hours), [Python](https://pypi.org/project/osm-humanized-opening-hours/), [Java](https://github.com/simonpoole/OpeningHoursParser).

**Comment by hb0nd** [comment]
There are also some premade implementations here: https://wiki.openstreetmap.org/wiki/Key:opening_hours#Implementation

Maybe it'd be better to use one of them?

**Reply by biodranik** [reply]
No, they won't work in our case for different reasons.

**Comment by biodranik** [comment]
Glad that you raised this question. The current implementation is _awful_ and has many issues (you can search on Github). I was strongly against using boost::spirit, but my opinion was rejected for political reasons 7 years ago.

The proper implementation in _Organic Maps_ style should be written in C++ (to be reused on all platforms), be lightweight, performant, robust, and support ALL valid combinations of opening hours, including months, years, and most importantly, Public Holidays. There are tons of ready unit tests, so it would be easy to check/validate the implementation.

Please let me know if you are interested, so we can design and implement it together.

**Comment by biodranik** [comment]
Let's start by collecting all existing Opening Hours issues into one meta-issue.
